File: app/src/main/java/com/github/bjoernpetersen/q/api/Permission.java
package com.github.bjoernpetersen.q.api;

import android.os.Parcel;
import android.os.Parcelable;

enum Permission implements Parcelable {
  SKIP("skip"), DISLIKE("dislike");

  private final String name;

  Permission(String name) {
    this.name = name;
  }

  public String getName() {
    return name;
  }

  public static Permission matchByName(String name) {
    for (Permission permission : values()) {
      if (permission.getName().equals(name)) {
        return permission;
      }
    }
    throw new IllegalArgumentException("Unknown permission: " + name);
  }

  private static Permission get(Parcel in) {
    String name = in.readString();
    return matchByName(name);
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    dest.writeString(name);
  }

  @Override
  public int describeContents() {
    return 0;
  }

  public static final Creator<Permission> CREATOR = new Creator<Permission>() {
    @Override
    public Permission createFromParcel(Parcel in) {
      return Permission.get(in);
    }

    @Override
    public Permission[] newArray(int size) {
      return new Permission[size];
    }
  };
}

File: app/src/main/java/com/github/bjoernpetersen/q/api/Permission.kt
package com.github.bjoernpetersen.q.api

import android.os.Parcel
import android.os.Parcelable

enum class Permission(val label: String) : Parcelable {
  SKIP("skip"), DISLIKE("dislike"), MOVE("move");

  override fun writeToParcel(dest: Parcel, flags: Int) = dest.writeString(name)

  override fun describeContents(): Int = 0

  companion object {
    @JvmStatic
    fun matchByLabel(label: String): Permission {
      values().filter { it.label == label }
          .forEach { return it }
      throw IllegalArgumentException("Unknown permission: " + label)
    }

    @JvmStatic
    private operator fun get(parcel: Parcel): Permission {
      val name = parcel.readString()
      return matchByLabel(name)
    }

    @JvmStatic
    val CREATOR: Parcelable.Creator<Permission> = object : Parcelable.Creator<Permission> {
      override fun createFromParcel(parcel: Parcel): Permission = Permission[parcel]
      override fun newArray(size: Int): Array<Permission?> = arrayOfNulls(size)
    }
  }
}
